### PR TITLE
Remove sorting in ArborX

### LINKED ISSUE #243: Disable sorting when calling ArborX
When calling ArborX inside the combination NimbleSM + BVH, the sorting of predicates in ArborX encounters sometimes the worst case for the bucket sorting algorithm.
This worst case yields large simulation times.

### PR DESCRIPTION
**Pull request by uhetmaniuk**
Closes #243 

Remove sorting in ArborX. 

Use Unmanaged Kokkos::View. 
Simplify computation of characteristic length.

### CODE REVIEW

File: src/contact/parallel/bvh_contact_manager.cc [hunk 1]
```diff
@@ -201,9 +208,12 @@ struct NarrowphaseFunc
     // Size of offset = n + 1
     //
     std::vector<NarrowphaseResult> resa_vec, resb_vec;
+    ArborX::Experimental::TraversalPolicy policy;
+    policy._sort_predicates = false;
```
Comment (dalg24) [inline src/contact/parallel/bvh_contact_manager.cc]:
```suggestion
    ArborX::Experimental::TraversalPolicy policy;
    policy.setPredicateSorting(false);
```

File: src/contact/parallel/bvh_contact_manager.cc [hunk 2]
```diff
@@ -201,9 +208,12 @@ struct NarrowphaseFunc
...
     //
-    a_bvh.query(nimble_kokkos::kokkos_device_execution_space{}, d_contact_nodes,
-              details::ArborXCallback{d_contact_faces, d_contact_nodes, contact_manager->GetPenaltyForceParam(), resa_vec, resb_vec});
+    a_bvh.query(nimble_kokkos::kokkos_host_execution_space{}, view_b,
+                details::ArborXCallback{view_a, view_b, contact_manager->GetPenaltyForceParam(), resa_vec, resb_vec},
+                policy);
```
Comment (dalg24) [inline src/contact/parallel/bvh_contact_manager.cc]: Alternatively,
```suggestion
                ArborX::Experimental::TraversalPolicy().setPredicateSorting(false));
```

Comment (uhetmaniuk) [reply src/contact/parallel/bvh_contact_manager.cc]: Thank you. I committed the first version.

**Comment by aprokop** [comment]
It is unclear to me that unconditionally disabling sorting is what you really want, as it is often faster on GPU. We know that `Kokkos::BinSort` is pretty terrible in many situations, and maybe this issue can be addressed by specializing sort routines on the host.

@uhetmaniuk Do you see a performance penalty on GPU when you disable sorting?

**Comment by uhetmaniuk** [comment]
@aprokop   We are not running on GPU for the moment.
What do you recommend when you hit a terrible case for `Kokkos::BinSort` at every iteration of your simulation (by terrible, I mean a query that takes 7s versus 0.5s for similar queries with sorting -- versus less 0.01s without sorting) ?
Of course this is one problem but we do sort our inputs for ArborX with a geometry-based sorting.

**Comment by aprokop** [comment]
@uhetmaniuk What backend are you running? If it's `Serial`, we could simply wrap `std::sort` for object sorting. However, if it's `OpenMP`, it becomes tricky, as we don't know of a great threaded sort, though there are some decent ones.

>  Of course this is one problem but we do sort our inputs for ArborX with a geometry-based sorting.

This may help.

**Comment by uhetmaniuk** [comment]
These numbers were with Serial backend.

**Comment by uhetmaniuk** [comment]
@aprokop I will send later today the set of primitives / predicates that were creating the 7s problem (per Damien's request). I have not visualized them yet but maybe they could be helpful.

**Comment by nmm0** [comment]
@dalg24 @aprokop The parallel driver is clustering the contact entities per rank. Could this be causing all of the bounding volumes to be placed into the same bin and thus degrade search? Is there a way to remap the bins to the range of the bounding volumes (we have the root bounding volume for the whole set of contact entities) with the ArborX API?

**Comment by aprokop** [comment]
> @dalg24 @aprokop The parallel driver is clustering the contact entities per rank. Could this be causing all of the bounding volumes to be placed into the same bin and thus degrade search? Is there a way to remap the bins to the range of the bounding volumes (we have the root bounding volume for the whole set of contact entities) with the ArborX API?

@nmm0 Sorry, I don't get what you are saying at all. Maybe you could elaborate some more on this?

I think what you are alluding to is whether close contact entities will same quantities associated with them (what you may call bins, but really Morton indices).

I think that's possible but unlikely. We could easily check this. Essentially, we could do `bvh.bounds()` to get the scene bounding box for the primitives, and then do
```c++
  // points
  auto const n = ponts.size();
  auto bbox = bvh.bounds();
  Kokkos::View<unsigned int *, MemorySpace> morton_indices(
      Kokkos::ViewAllocateWithoutInitializing("morton"), n);
  Details::TreeConstruction::assignMortonCodes(space, points, morton_indices, bbox);
  auto permute = ArborX::Details::sortObjects(morton_indices);
```
and then count the number of repeated Morton indices.

**Review by nmm0 — APPROVED** [review]
Looks good to me!

**Comment by uhetmaniuk** [comment]
@nmm0
This version requires a patch for ArborX where the sorting with `Kokkos::BinSort` is replaced with `std::sort`.
This patch is useful when Kokkos is used in Serial mode and the bounding box for all the faces does not include all the potential contact nodes (i.e. a majority of potential contact nodes are outside of the bounding box).

**Comment by aprokop** [comment]
It may take some time in ArborX to fix this, as `std::sort` is not unconditionally faster than `Kokkos::BinSort`, so we may need to figure out the heuristics.

**Pull request merged**